receiver/mongodbreceiver/extractor.go

```diff
@@ -0,0 +1,247 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
+)
+
+type extractor struct {
+	version *version.Version
+	logger  *zap.Logger
+}
+
+type extractType int
+
+const (
+	adminServerStats extractType = iota
+	normalDBStats
+	normalServerStats
+)
+
+// Mongo26 is a version representation of mongo 2.6
+var Mongo26, _ = version.NewVersion("2.6")
+
+// Mongo30 is a version representation of mongo 3.0
+var Mongo30, _ = version.NewVersion("3.0")
+
+// Mongo40 is a version representation of mongo 4.0
+var Mongo40, _ = version.NewVersion("4.0")
+
+// Mongo42 is a version representation of mongo 4.2
+var Mongo42, _ = version.NewVersion("4.2")
+
+// Mongo44 is a version representation of mongo 4.4
+var Mongo44, _ = version.NewVersion("4.4")
+
+// Mongo50 is a version representation of mongo 5.0
+var Mongo50, _ = version.NewVersion("5.0")
+
+// newExtractor returns a representation for a utility to dig through arbitrary bson
+// for metric values
+func newExtractor(mongoVersion string, logger *zap.Logger) (*extractor, error) {
+	v, err := version.NewVersion(mongoVersion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse version as semantic version %s: %w", mongoVersion, err)
+	}
+	return &extractor{
+		version: v,
+		logger:  logger,
+	}, nil
+}
+
+func (e *extractor) Extract(document bson.M, mb *metadata.MetricsBuilder, dbName string, et extractType) {
+	now := pdata.NewTimestampFromTime(time.Now())
+	switch et {
+	case normalDBStats:
+		e.extractDBStats(now, document, mb, dbName)
+	case normalServerStats:
+		e.extractNormalServerStats(now, document, mb, dbName)
+	case adminServerStats:
+		e.extractAdminStats(now, document, mb)
+	}
+}
+
+func (e *extractor) extractDBStats(ts pdata.Timestamp, doc bson.M, mb *metadata.MetricsBuilder, dbName string) {
+	collectionsPath := []string{"collections"}
+	if value, err := digForIntValue(doc, collectionsPath); err == nil {
```

Comment:
We need to aggregate errors using the `scrapererror` package. [Here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/couchdbreceiver/metrics.go) is a good example of it being used in a similar situation.